File: tests/integration/api_client.py
"""
BharatBuild AI - API Client for Integration Testing
Direct API calls to test backend endpoints.
"""
import aiohttp
import asyncio
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
import json

from config import config


@dataclass
class APIResponse:
    """API Response wrapper"""
    status: int
    data: Any
    headers: Dict[str, str]
    success: bool

    @property
    def json(self) -> Any:
        return self.data


class BharatBuildAPIClient:
    """API Client for BharatBuild AI backend"""

    def __init__(self, base_url: str = None):
        self.base_url = (base_url or config.api_base_url).rstrip('/')
        self.token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        self.session: Optional[aiohttp.ClientSession] = None

    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()

    def _get_headers(self, auth: bool = True) -> Dict[str, str]:
        """Get request headers"""
        headers = {"Content-Type": "application/json"}
        if auth and self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        return headers

    async def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        auth: bool = True,
        timeout: int = None
    ) -> APIResponse:
        """Make HTTP request"""
        url = f"{self.base_url}{endpoint}"
        headers = self._get_headers(auth)
        timeout_obj = aiohttp.ClientTimeout(total=timeout or config.request_timeout)

        try:
            async with self.session.request(
                method, url, json=data, headers=headers, timeout=timeout_obj
            ) as response:
                try:
                    response_data = await response.json()
                except:
                    response_data = await response.text()

                return APIResponse(
                    status=response.status,
                    data=response_data,
                    headers=dict(response.headers),
                    success=200 <= response.status < 300
                )
        except Exception as e:
            return APIResponse(
                status=0,
                data={"error": str(e)},
                headers={},
                success=False
            )

    # ==================== Health ====================
    async def health_check(self) -> APIResponse:
        """Check API health"""
        return await self._request("GET", "/health", auth=False)

    # ==================== Authentication ====================
    async def register(
        self,
        email: str,
        password: str,
        full_name: str,
        role: str = "student",
        **kwargs
    ) -> APIResponse:
        """Register a new user"""
        data = {
            "email": email,
            "password": password,
            "full_name": full_name,
            "role": role,
            "roll_number": kwargs.get("roll_number", "TEST001"),
            "college_name": kwargs.get("college_name", "Test College"),
            "university_name": kwargs.get("university_name", "Test University"),
            "department": kwargs.get("department", "Computer Science"),
            "course": kwargs.get("course", "B.Tech"),
            "year_semester": kwargs.get("year_semester", "4th Year"),
            "batch": kwargs.get("batch", "2021-2025"),
            "guide_name": kwargs.get("guide_name", "Dr. Test Guide"),
            "guide_designation": kwargs.get("guide_designation", "Professor"),
            "hod_name": kwargs.get("hod_name", "Dr. HOD Test")
        }
        return await self._request("POST", "/auth/register", data=data, auth=False)

    async def login(self, email: str, password: str) -> APIResponse:
        """Login user"""
        data = {"email": email, "password": password}
        response = await self._request("POST", "/auth/login", data=data, auth=False)

        if response.success and isinstance(response.data, dict):
            self.token = response.data.get("access_token")
            self.refresh_token = response.data.get("refresh_token")

        return response

    async def get_current_user(self) -> APIResponse:
        """Get current user info"""
        return await self._request("GET", "/auth/me")

    async def refresh_access_token(self) -> APIResponse:
        """Refresh access token"""
        data = {"refresh_token": self.refresh_token}
        response = await self._request("POST", "/auth/refresh", data=data, auth=False)

        if response.success and isinstance(response.data, dict):
            self.token = response.data.get("access_token")

        return response

    async def logout(self) -> APIResponse:
        """Logout user"""
        response = await self._request("POST", "/auth/logout")
        self.token = None
        self.refresh_token = None
        return response

    # ==================== Projects ====================
    async def list_projects(self, page: int = 1, page_size: int = 10) -> APIResponse:
        """List user projects"""
        return await self._request("GET", f"/projects?page={page}&page_size={page_size}")

    async def get_project(self, project_id: str) -> APIResponse:
        """Get project details"""
        return await self._request("GET", f"/projects/{project_id}")

    async def create_project(self, name: str, prompt: str, project_type: str = "web") -> APIResponse:
        """Create a new project"""
        data = {
            "name": name,
            "prompt": prompt,
            "project_type": project_type
        }
        return await self._request("POST", "/projects", data=data)

    async def delete_project(self, project_id: str) -> APIResponse:
        """Delete a project"""
        return await self._request("DELETE", f"/projects/{project_id}")

    async def get_project_files(self, project_id: str) -> APIResponse:
        """Get project files"""
        return await self._request("GET", f"/projects/{project_id}/files")

    async def get_file_content(self, project_id: str, file_path: str) -> APIResponse:
        """Get file content"""
        return await self._request("GET", f"/projects/{project_id}/files/{file_path}")

    # ==================== Documents ====================
    async def get_document_types(self) -> APIResponse:
        """Get available document types"""
        return await self._request("GET", "/documents/types", auth=False)

    async def list_documents(self, project_id: str) -> APIResponse:
        """List project documents"""
        return await self._request("GET", f"/documents/list/{project_id}")

    async def generate_document(self, project_id: str, doc_type: str) -> APIResponse:
        """Generate a document"""
        return await self._request("POST", f"/documents/generate/{project_id}/{doc_type}", timeout=120)

    async def download_document(self, project_id: str, doc_type: str) -> APIResponse:
        """Download a document"""
        return await self._request("GET", f"/documents/download/{project_id}/{doc_type}")

    # ==================== Users ====================
    async def get_plan_status(self) -> APIResponse:
        """Get user's plan status"""
        return await self._request("GET", "/users/plan-status")

    async def get_token_balance(self) -> APIResponse:
        """Get user's token balance"""
        return await self._request("GET", "/users/token-balance")

    async def update_profile(self, data: Dict) -> APIResponse:
        """Update user profile"""
        return await self._request("PUT", "/users/profile", data=data)

    # ==================== Execution ====================
    async def run_project(self, project_id: str) -> APIResponse:
        """Run/execute a project"""
        return await self._request("POST", f"/execution/run/{project_id}", timeout=60)

    async def stop_execution(self, project_id: str) -> APIResponse:
        """Stop project execution"""
        return await self._request("POST", f"/execution/stop/{project_id}")

    async def export_project(self, project_id: str) -> APIResponse:
        """Export project as ZIP"""
        return await self._request("GET", f"/execution/export/{project_id}")


# Synchronous wrapper for non-async contexts
class SyncAPIClient:
    """Synchronous wrapper for API client"""

    def __init__(self, base_url: str = None):
        self.base_url = base_url

    def _run(self, coro):
        """Run coroutine synchronously"""
        loop = asyncio.new_event_loop()
        try:
            return loop.run_until_complete(coro)
        finally:
            loop.close()

    def health_check(self):
        async def _call():
            async with BharatBuildAPIClient(self.base_url) as client:
                return await client.health_check()
        return self._run(_call())

    def login(self, email: str, password: str):
        async def _call():
            async with BharatBuildAPIClient(self.base_url) as client:
                return await client.login(email, password)
        return self._run(_call())

    def register(self, email: str, password: str, full_name: str, **kwargs):
        async def _call():
            async with BharatBuildAPIClient(self.base_url) as client:
                return await client.register(email, password, full_name, **kwargs)
        return self._run(_call())


if __name__ == "__main__":
    # Quick test
    async def test():
        async with BharatBuildAPIClient() as client:
            # Health check
            response = await client.health_check()
            print(f"Health: {response.status} - {response.data}")

            # Document types
            response = await client.get_document_types()
            print(f"Doc Types: {response.status} - {len(response.data) if isinstance(response.data, list) else response.data}")

    asyncio.run(test())

File: frontend/src/lib/api-client.ts
import axios, { AxiosInstance } from 'axios'
import { setAccessToken, removeAccessToken } from './auth-utils'

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1'

/**
 * Extended API Client with all BharatBuild methods
 */
class ApiClient {
  private axiosInstance: AxiosInstance

  constructor() {
    this.axiosInstance = axios.create({
      baseURL: API_BASE_URL,
      headers: {
        'Content-Type': 'application/json',
      },
      timeout: 180000, // 3 minutes - SDK Fixer Agent needs more time for complex fixes
    })

    // Add auth token to requests
    this.axiosInstance.interceptors.request.use(
      (config) => {
        const token = typeof window !== 'undefined' ? localStorage.getItem('access_token') : null
        if (token) {
          config.headers.Authorization = `Bearer ${token}`
        }
        return config
      },
      (error) => Promise.reject(error)
    )

    // Handle response errors
    this.axiosInstance.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 401) {
          if (typeof window !== 'undefined') {
            // Don't redirect if already on login page
            if (window.location.pathname === '/login') {
              return Promise.reject(error)
            }

            // Try to refresh token first before logging out
            const refreshToken = localStorage.getItem('refresh_token')
            if (refreshToken && !error.config._retry) {
              error.config._retry = true
              try {
                const response = await axios.post(
                  `${API_BASE_URL}/auth/refresh`,
                  { refresh_token: refreshToken },
                  { headers: { 'Content-Type': 'application/json' } }
                )
                if (response.data?.access_token) {
                  // Update token in both localStorage and cookie
                  setAccessToken(response.data.access_token)
                  // Retry the original request with new token
                  error.config.headers.Authorization = `Bearer ${response.data.access_token}`
                  return this.axiosInstance.request(error.config)
                }
              } catch (refreshError) {
                // Refresh failed, just reject - let pages handle gracefully
                console.log('Token refresh failed')
              }
            }

            // Don't auto-redirect to login - let pages handle 401 gracefully
            // Pages can show mock data or handle auth errors as needed
            // Only clear tokens if refresh failed
            if (error.config._retry) {
              removeAccessToken()
              localStorage.removeItem('refresh_token')
            }
          }
        }

        // Handle feature restriction errors (403 with feature_not_available)
        if (error.response?.status === 403) {
          const detail = error.response?.data?.detail
          if (detail?.error === 'feature_not_available' && typeof window !== 'undefined') {
            // Emit custom event for UpgradeContext to catch
            const event = new CustomEvent('feature-restricted', {
              detail: {
                feature: detail.feature,
                currentPlan: detail.current_plan,
                upgradeTo: detail.upgrade_to,
                message: detail.message
              }
            })
            window.dispatchEvent(event)
          }
        }

        return Promise.reject(error)
      }
    )
  }

  // ==================== Auth ====================
  async login(email: string, password: string) {
    const response = await this.axiosInstance.post('/auth/login', { email, password })
    return response.data
  }

  async register(data: {
    email: string;
    password: string;
    full_name: string;
    role?: string;
    phone?: string;
    // Student Academic Details
    roll_number?: string;
    college_name?: string;
    university_name?: string;
    department?: string;
    course?: string;
    year_semester?: string;
    batch?: string;
    // Guide Details
    guide_name?: string;
    guide_designation?: string;
    hod_name?: string;
    // Legacy fields
    college?: string;
    year_of_study?: string;
    faculty_id?: string;
    designation?: string;
  }) {
    const response = await this.axiosInstance.post('/auth/register', data)
    return response.data
  }

  async logout() {
    const response = await this.axiosInstance.post('/auth/logout')
    return response.data
  }

  async getMe() {
    const response = await this.axiosInstance.get('/auth/me')
    return response.data
  }

  async initiateOAuth(provider: 'google' | 'github', role?: string) {
    const response = await this.axiosInstance.get(`/auth/${provider}/authorize`, {
      params: { role }
    })
    if (response.data?.authorization_url) {
      window.location.href = response.data.authorization_url
    }
    return response.data
  }

  async googleCallback(code: string, role?: string) {
    const response = await this.axiosInstance.post('/auth/google/callback', { code, role })
    return response.data
  }

  async githubCallback(code: string, role?: string) {
    const response = await this.axiosInstance.post('/auth/github/callback', { code, role })
    return response.data
  }

  // ==================== Health ====================
  async healthCheck() {
    try {
      const response = await this.axiosInstance.get('/health', { baseURL: API_BASE_URL.replace('/api/v1', '') })
      return response.data?.status === 'healthy'
    } catch {
      return false
    }
  }

  // ==================== Projects ====================
  async getProjects(params?: { page?: number; limit?: number }) {
    // Backend expects page_size, not limit
    const queryParams = params ? { page: params.page, page_size: params.limit } : undefined
    const response = await this.axiosInstance.get('/projects', { params: queryParams })
    return response.data
  }

  async getProject(id: string) {
    const response = await this.axiosInstance.get(`/projects/${id}`)
    return response.data
  }

  async createProject(data: {
    name?: string;
    title?: string;
    description?: string;
    type?: string;
    prompt?: string;
    mode?: string;
    tech_stack?: string;
    features?: string[];
  }) {
    const response = await this.axiosInstance.post('/projects', data)
    return response.data
  }

  async updateProject(id: string, data: Partial<{ name: string; description: string }>) {
    const response = await this.axiosInstance.patch(`/projects/${id}`, data)
    return response.data
  }

  async deleteProject(id: string) {
    const response = await this.axiosInstance.delete(`/projects/${id}`)
    return response.data
  }

  async executeProject(id: string) {
    const response = await this.axiosInstance.post(`/execution/${id}/start`)
    return response.data
  }

  async getProjectStatus(id: string) {
    const response = await this.axiosInstance.get(`/execution/${id}/status`)
    return response.data
  }

  async getProjectFiles(id: string) {
    const response = await this.axiosInstance.get(`/projects/${id}/files`)
    return response.data
  }

  async getFileContent(projectId: string, filePath: string) {
    const response = await this.axiosInstance.get(`/projects/${projectId}/files/${encodeURIComponent(filePath)}`)
    return response.data
  }

  async loadProjectWithFiles(projectId: string) {
    // Uses the /load endpoint which returns project + all files with content
    const response = await this.axiosInstance.get(`/projects/${projectId}/load`)
    return response.data
  }

  async saveFile(projectId: string, file: { path: string; content: string }) {
    const response = await this.axiosInstance.post('/sync/file', {
      project_id: projectId,
      path: file.path,
      content: file.content,
    })
    return response.data
  }

  async saveFilesBulk(projectId: string, files: Array<{ path: string; content: string }>) {
    const response = await this.axiosInstance.post(`/sync/files/${projectId}`, { files })
    return response.data
  }

  // ==================== Documents ====================
  async downloadDocument(projectId: string, docType: string) {
    const response = await this.axiosInstance.get(`/documents/${projectId}/${docType}`, {
      responseType: 'blob'
    })
    return response.data
  }

  // ==================== Plan Status ====================
  async getPlanStatus() {
    const response = await this.axiosInstance.get('/billing/status')
    return response.data
  }

  // ==================== Tokens ====================
  async getTokenBalance() {
    const response = await this.axiosInstance.get('/tokens/balance')
    return response.data
  }

  async getTokenPackages() {
    const response = await this.axiosInstance.get('/tokens/packages')
    return response.data
  }

  async purchaseTokens(packageId: string) {
    const response = await this.axiosInstance.post('/tokens/purchase', { package_id: packageId })
    return response.data
  }

  async redeemPromoCode(code: string) {
    const response = await this.axiosInstance.post('/tokens/redeem', { code })
    return response.data
  }

  async getTokenAnalytics() {
    const response = await this.axiosInstance.get('/tokens/analytics')
    return response.data
  }

  async getTokenHistory(params?: { page?: number; limit?: number }) {
    const response = await this.axiosInstance.get('/tokens/history', { params })
    return response.data
  }

  // ==================== Orchestrator ====================
  /**
   * Cancel an ongoing project generation
   * Stops file generation on the backend and frees up resources
   */
  async cancelGeneration(projectId: string) {
    const response = await this.axiosInstance.post('/orchestrator/cancel', { project_id: projectId })
    return response.data
  }

  /**
   * Get project generation status (for polling when SSE disconnects)
   */
  async getGenerationStatus(projectId: string) {
    const response = await this.axiosInstance.get(`/orchestrator/project/${projectId}/status`)
    return response.data
  }

  /**
   * Get detailed generation progress with file statuses
   */
  async getGenerationProgress(projectId: string) {
    const response = await this.axiosInstance.get(`/orchestrator/project/${projectId}/progress`)
    return response.data
  }

  /**
   * Resume an interrupted project generation
   */
  async resumeGeneration(projectId: string, continueMessage?: string) {
    const response = await this.axiosInstance.post('/orchestrator/resume', {
      project_id: projectId,
      continue_message: continueMessage || 'Continue generating the remaining files'
    })
    return response.data
  }

  // ==================== APK Downloads (Flutter/Mobile) ====================

  /**
   * Get APK availability info for a project
   * Returns whether APK is available and download URL
   */
  async getApkInfo(projectId: string) {
    const response = await this.axiosInstance.get(`/projects/${projectId}/download/apk/info`)
    return response.data
  }

  /**
   * Get APK info for a session-based project (ephemeral storage)
   */
  async getSessionApkInfo(sessionId: string) {
    const response = await this.axiosInstance.get(`/download/session/${sessionId}/apk/info`)
    return response.data
  }

  /**
   * Download APK file for a project
   * Returns blob for download
   */
  async downloadApk(projectId: string, buildType: 'release' | 'debug' = 'release') {
    const response = await this.axiosInstance.get(`/projects/${projectId}/download/apk`, {
      params: { build_type: buildType },
      responseType: 'blob'
    })
    return response.data
  }

  /**
   * Download APK file for a session-based project
   */
  async downloadSessionApk(sessionId: string, buildType: 'release' | 'debug' = 'release') {
    const response = await this.axiosInstance.get(`/download/session/${sessionId}/apk`, {
      params: { build_type: buildType },
      responseType: 'blob'
    })
    return response.data
  }

  /**
   * Trigger APK build for a Flutter project
   */
  async triggerApkBuild(sessionId: string, buildType: 'release' | 'debug' = 'debug') {
    const response = await this.axiosInstance.post(`/download/session/${sessionId}/build-apk`, null, {
      params: { build_type: buildType }
    })
    return response.data
  }

  // ==================== Generic request methods ====================
  async get<T = any>(url: string, config?: any): Promise<T> {
    const response = await this.axiosInstance.get(url, config)
    return response.data
  }

  async post<T = any>(url: string, data?: any, config?: any): Promise<T> {
    const response = await this.axiosInstance.post(url, data, config)
    return response.data
  }

  async put<T = any>(url: string, data?: any, config?: any): Promise<T> {
    const response = await this.axiosInstance.put(url, data, config)
    return response.data
  }

  async patch<T = any>(url: string, data?: any, config?: any): Promise<T> {
    const response = await this.axiosInstance.patch(url, data, config)
    return response.data
  }

  async delete<T = any>(url: string, config?: any): Promise<T> {
    const response = await this.axiosInstance.delete(url, config)
    return response.data
  }

  // ==================== Lab Management (Faculty) ====================

  // Get faculty's labs
  async getFacultyLabs() {
    const response = await this.axiosInstance.get('/faculty/labs')
    return response.data
  }

  // Get lab details
  async getLab(labId: string) {
    const response = await this.axiosInstance.get(`/lab/labs/${labId}`)
    return response.data
  }

  // Get lab topics (experiments)
  async getLabTopics(labId: string) {
    const response = await this.axiosInstance.get(`/lab/labs/${labId}/topics`)
    return response.data
  }

  // Create lab topic (experiment)
  async createLabTopic(labId: string, data: {
    title: string
    description?: string
    week_number?: number
    concept_content?: string
    prerequisites?: string[]
  }) {
    const response = await this.axiosInstance.post(`/lab/labs/${labId}/topics`, data)
    return response.data
  }

  // Update lab topic
  async updateLabTopic(topicId: string, data: any) {
    const response = await this.axiosInstance.put(`/lab/topics/${topicId}`, data)
    return response.data
  }

  // Get topic MCQs
  async getTopicMCQs(topicId: string) {
    const response = await this.axiosInstance.get(`/lab/topics/${topicId}/mcqs`)
    return response.data
  }

  // Create MCQ
  async createMCQ(topicId: string, data: {
    question_text: string
    options: string[]
    correct_option: number
    explanation?: string
    difficulty?: string
    marks?: number
  }) {
    const response = await this.axiosInstance.post(`/lab/topics/${topicId}/mcqs`, data)
    return response.data
  }

  // Get topic coding problems
  async getTopicProblems(topicId: string) {
    const response = await this.axiosInstance.get(`/lab/topics/${topicId}/problems`)
    return response.data
  }

  // Create coding problem
  async createCodingProblem(topicId: string, data: {
    title: string
    description: string
    difficulty?: string
    max_score?: number
    supported_languages?: string[]
    starter_code?: Record<string, string>
    test_cases?: Array<{ input: string; expected_output: string; is_hidden?: boolean }>
  }) {
    const response = await this.axiosInstance.post(`/lab/topics/${topicId}/problems`, data)
    return response.data
  }

  // Get lab students with progress
  async getLabStudents(labId: string) {
    const response = await this.axiosInstance.get(`/lab/labs/${labId}/students`)
    return response.data
  }

  // Get lab analytics
  async getLabAnalytics(labId: string) {
    const response = await this.axiosInstance.get(`/lab/labs/${labId}/analytics`)
    return response.data
  }

  // Get faculty classes/sections
  async getFacultyClasses() {
    const response = await this.axiosInstance.get('/faculty/classes')
    return response.data
  }

  // Assign topic to class/batch
  async assignTopicToClass(data: {
    topic_id: string
    lab_id: string
    class_id?: string
    batch_id?: string
    student_ids?: string[]
    deadline: string
    deadline_type: 'soft' | 'hard'
  }) {
    const response = await this.axiosInstance.post('/faculty/lab-assignments', data)
    return response.data
  }

  // Get lab assignments
  async getLabAssignments(labId: string) {
    const response = await this.axiosInstance.get(`/faculty/lab-assignments/${labId}`)
    return response.data
  }

  // Update lab assignment
  async updateLabAssignment(assignmentId: string, data: any) {
    const response = await this.axiosInstance.put(`/faculty/lab-assignments/${assignmentId}`, data)
    return response.data
  }

  // Get student progress for a specific topic
  async getTopicStudentProgress(topicId: string) {
    const response = await this.axiosInstance.get(`/lab/topics/${topicId}/student-progress`)
    return response.data
  }
}

export const apiClient = new ApiClient()
export default apiClient
